### ISSUE TITLE
KeyError: 6

### ISSUE THREAD

**Issue by tinyapps** [relikd/abcddb2vcard#1]
Thanks so much for crafting and sharing abcddb2vcard, relikd.

When running it against an AddressBook-v22.abcddb file with some contacts that refused to export as vCard from Contacts.app, it returned the following error:

```
% python3 abcddb2vcard.py contacts_$(date +"%Y-%m-%d").vcf
Traceback (most recent call last):
  File "/Users/user/Desktop/abcddb2vcard-main/abcddb2vcard.py", line 15, in <module>
    contacts = ABCDDB.load(Path.home().joinpath(
  File "/Users/user/Desktop/abcddb2vcard-main/ABCDDB.py", line 21, in load
    records[x.parent].email.append(x)
KeyError: 6
```

Dragging the All Contacts card from Contacts.app to the Desktop exported all but the affected contacts to "All Contacts.vcf".

I was able to export all of the contacts to vCard, including the ones that refused to export otherwise, by using Automator as described [here](https://tinyapps.org/blog/202302190700_contacts_empty_vcard.html).

**Comment by relikd** [comment]
ohh, interesting. :) ... it seems there is an email address pointing to a contact that does not exist. Don't post it here (for your own privacy) but does the following query returns anything?

```sh
sqlite3 "$HOME/Library/Application Support/AddressBook/AddressBook-v22.abcddb" "SELECT ZFIRSTNAME, ZLASTNAME FROM ZABCDRECORD WHERE Z_PK = 6"
```

If the query above indeed returns nothing, you can check whether your Automator workflow did export the email address in question:

```sh
 sqlite3 "$HOME/Library/Application Support/AddressBook/AddressBook-v22.abcddb" "SELECT ZADDRESS FROM ZABCDEMAILADDRESS WHERE ZOWNER = 6"
```

I wonder what happend. I cannot ask you for your address book, but can you reproduce the bug in a blank abcddb?

**Comment by tinyapps** [comment]
Thanks so much for the fast reply!

Yes, that does return the first and last name of that contact, which is the first of 77 affected contacts (i.e., ones that won't export to vCard from Contacts.app).

In fact, the most interesting table in the database, ZABCDCONTACTINDEX, displays all of the information for all of the contacts, including the affected ones:

`select * from ZABCDCONTACTINDEX;`

It seems that tools which access AddressBook-v22.abcddb like Contacts.app, abcddb2vcard, and [Exporter for Contacts 2](https://www.subclassed.com/apps/exporter-for-contacts-2/export-mac-contacts-to-csv-or-excel/index.html) fail on the affected contacts (despite sqlite3 queries working from Terminal) while tools which access the associated .abcdp files like [contact_reaper.py](https://gist.github.com/andermoran/7551e17bccd334b2f304848520b41aaf) and Automator are able to export the affected contacts.

I'm sorry that I cannot share the AddressBook-v22.abcddb file with you as it doesn't belong to me but to a friend.

**Comment by relikd** [comment]
> Yes, that does return the first and last name of that contact

Well, that is weird. Because if the ID exists, python should not throw a KeyError.

I've pushed some changes. This should prevent the tool from crashing but will export the data into a blank contact. So you will loose the name of the contact. Can you report back how this is broken compared to the expected? thanks

**Comment by tinyapps** [comment]
Thank you very much. Here is the new error message:

```
Traceback (most recent call last):
  File "/Users/user/Desktop/abcddb2vcard-main/abcddb2vcard.py", line 15, in <module>
    contacts = ABCDDB.load(Path.home().joinpath(
  File "/Users/user/Desktop/abcddb2vcard-main/ABCDDB.py", line 422, in load
    records = Record.queryAll(cur)
  File "/Users/user/Desktop/abcddb2vcard-main/ABCDDB.py", line 284, in queryAll
    return {x[0]: Record(x) for x in cursor.execute('''
  File "/Users/user/Desktop/abcddb2vcard-main/ABCDDB.py", line 284, in <dictcomp>
    return {x[0]: Record(x) for x in cursor.execute('''
  File "/Users/user/Desktop/abcddb2vcard-main/ABCDDB.py", line 324, in __init__
    self.iscompany = row[17] & 1  # type: bool
TypeError: unsupported operand type(s) for &: 'NoneType' and 'int'
```

**Comment by relikd** [comment]
and what about now?

**Comment by tinyapps** [comment]
Awesome! Now the Terminal output shows `276 contacts` (which is the same number appearing in Contacts.app, including both affected and unaffected contacts). The created .vcf file was reported by Quick Look to contain 252 contacts (which includes a large number of the 77 affected contacts), and importing the .vcf file into a blank Contacts.app resulted in 252 contacts being imported.

A quick look at the .vcf file in a text editor shows the 24 missing contacts inside with the other 252 contacts; some of their details have been merged into other contacts after import into Contacts.app. I'll have a closer look for any similarities between those 24 and post back here later today.

Thanks so much!

**Comment by tinyapps** [comment]
Great news, relikd! The missing 24 contacts in Quick Look and Contacts.app were caused by a bug in macOS 13.0.1's handling of vCard files. After upgrading to 13.2.1, both Quick Look and Contacts.app correctly displayed all 276 contacts which abcddb2vcard exported from AddressBook-v22.abcddb!

The only reason I stumbled onto this was because the virtual machine where I was testing was running 13.0.1; when I copied the vCard file to the host machine running 13.2.1, Quick Look reported the correct number, 276 (just as abcddb2vcard's Terminal output had done). I double checked Quick Look in 13.0.1; it still showed 252. After upgrading the VM to 13.2.1, Quick Look displayed 276, as did Contacts.app after importing them.

Thank you again for such a swift resolution!
